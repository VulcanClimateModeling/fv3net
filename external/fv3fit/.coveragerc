[run]
omit = tests/* setup.py fv3fit/sklearn/__main__.py fv3fit/keras/__main__.py
