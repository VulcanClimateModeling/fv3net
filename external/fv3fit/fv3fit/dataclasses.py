from dataclasses import asdict
from enum import Enum


def asdict_with_enum(obj):
    """Recursively turn a dataclass obj into a dictionary handling any Enums
    """

    def _generate(x):
        for key, val in x:
            if isinstance(val, Enum):
                yield key, val.value
            else:
                yield key, val

    def dict_factory(x):
        return dict(_generate(x))

    return asdict(obj, dict_factory=dict_factory)
