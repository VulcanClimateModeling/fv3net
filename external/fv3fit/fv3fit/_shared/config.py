import dataclasses
from typing_extensions import Literal
import fsspec
import yaml
import os
from typing import (
    Any,
    Callable,
    Mapping,
    Optional,
    Tuple,
    Union,
    Sequence,
    List,
    Type,
    Dict,
)
from fv3fit.typing import Dataclass
import xarray as xr
from .predictor import Predictor
import dacite
import numpy as np
import random
import loaders

# TODO: move all keras configs under fv3fit.keras
import tensorflow as tf


DELP = "pressure_thickness_of_atmospheric_layer"
MODEL_CONFIG_FILENAME = "training_config.yml"


TrainingFunction = Callable[
    [
        Sequence[str],
        Sequence[str],
        Dataclass,
        Sequence[xr.Dataset],
        Sequence[xr.Dataset],
    ],
    Predictor,
]


def set_random_seed(seed: Union[float, int] = 0):
    # https://stackoverflow.com/questions/32419510/how-to-get-reproducible-results-in-keras
    os.environ["PYTHONHASHSEED"] = str(seed)
    np.random.seed(seed + 1)
    random.seed(seed + 2)
    tf.random.set_seed(seed + 3)


# TODO: delete this routine by refactoring the tests to no longer depend on it
def get_keras_model(name):
    return TRAINING_FUNCTIONS[name][0]


@dataclasses.dataclass
class TrainingConfig:
    """Convenience wrapper for model training parameters and file info

    Attrs:
        model_type: sklearn model type or keras model class to initialize
        input_variables: variables used as features
        output_variables: variables to predict
        hyperparameters: model_type-specific training configuration
        additional_variables: variables needed for training which aren't input
            or output variables of the trained model
        sample_dim_name: deprecated, internal name used for sample dimension
            when training and predicting
        random_seed: value to use to initialize randomness
    """

    model_type: str
    input_variables: List[str]
    output_variables: List[str]
    hyperparameters: Dataclass
    additional_variables: List[str] = dataclasses.field(default_factory=list)
    sample_dim_name: str = "sample"
    random_seed: Union[float, int] = 0

    @classmethod
    def from_dict(cls, kwargs) -> "TrainingConfig":
        kwargs = {**kwargs}  # make a copy to avoid mutating the input
        hyperparameter_class = get_hyperparameter_class(kwargs["model_type"])
        kwargs["hyperparameters"] = dacite.from_dict(
            data_class=hyperparameter_class, data=kwargs.get("hyperparameters", {})
        )
        return dacite.from_dict(data_class=cls, data=kwargs)


TRAINING_FUNCTIONS: Dict[str, Tuple[TrainingFunction, Type[Dataclass]]] = {}


def get_hyperparameter_class(model_type: str) -> Type:
    if model_type in TRAINING_FUNCTIONS:
        _, subclass = TRAINING_FUNCTIONS[model_type]
    else:
        raise ValueError(f"unknown model_type {model_type}")
    return subclass


def get_training_function(model_type: str) -> TrainingFunction:
    if model_type in TRAINING_FUNCTIONS:
        estimator_class, _ = TRAINING_FUNCTIONS[model_type]
    else:
        raise ValueError(f"unknown model_type {model_type}")
    return estimator_class


def register_training_function(name: str, hyperparameter_class: type):
    """
    Returns a decorator that will register the given training function
    to be usable in training configuration.
    """

    def decorator(func: TrainingFunction) -> TrainingFunction:
        TRAINING_FUNCTIONS[name] = (func, hyperparameter_class)
        return func

    return decorator


@dataclasses.dataclass
class OptimizerConfig:
    name: str
    kwargs: Mapping[str, Any] = dataclasses.field(default_factory=dict)

    @property
    def instance(self) -> tf.keras.optimizers.Optimizer:
        cls = getattr(tf.keras.optimizers, self.name)
        return cls(**self.kwargs)


@dataclasses.dataclass
class RegularizerConfig:
    name: str
    kwargs: Mapping[str, Any] = dataclasses.field(default_factory=dict)

    @property
    def instance(self) -> tf.keras.regularizers.Regularizer:
        cls = getattr(tf.keras.regularizers, self.name)
        return cls(**self.kwargs)


# TODO: move this class to where the Dense training is defined when config.py
# no longer depends on it (i.e. when _ModelTrainingConfig is deleted)
@dataclasses.dataclass
class DenseHyperparameters:
    """
    Configuration for training a dense neural network based model.

    Args:
        weights: loss function weights, defined as a dict whose keys are
            variable names and values are either a scalar referring to the total
            weight of the variable. Default is a total weight of 1
            for each variable.
        normalize_loss: if True (default), normalize outputs by their standard
            deviation before computing the loss function
        optimizer_config: selection of algorithm to be used in gradient descent
        kernel_regularizer_config: selection of regularizer for hidden dense layer
            weights, by default no regularization is applied
        depth: number of dense layers to use between the input and output layer.
            The number of hidden layers will be (depth - 1)
        width: number of neurons to use on layers between the input and output layer
        gaussian_noise: how much gaussian noise to add before each Dense layer,
            apart from the output layer
        loss: loss function to use, should be 'mse' or 'mae'
        spectral_normalization: whether to apply spectral normalization to hidden layers
        save_model_checkpoints: if True, save one model per epoch when
            dumping, under a 'model_checkpoints' subdirectory
        nonnegative_outputs: if True, add a ReLU activation layer as the last layer
            after output denormalization layer to ensure outputs are always >=0
            Defaults to False.
        fit_kwargs: other keyword arguments to be passed to the underlying
            tf.keras.Model.fit() method
    """

    weights: Optional[Mapping[str, Union[int, float]]] = None
    normalize_loss: bool = True
    optimizer_config: OptimizerConfig = dataclasses.field(
        default_factory=lambda: OptimizerConfig("Adam")
    )
    kernel_regularizer_config: Optional[RegularizerConfig] = None
    depth: int = 3
    width: int = 16
    epochs: int = 3
    gaussian_noise: float = 0.0
    loss: str = "mse"
    spectral_normalization: bool = False
    save_model_checkpoints: bool = False
    nonnegative_outputs: bool = False

    # TODO: remove fit_kwargs by fixing how validation data is passed
    fit_kwargs: Optional[dict] = None


@dataclasses.dataclass
class RandomForestHyperparameters:
    """
    Configuration for training a random forest based model.

    Trains one random forest for each training batch provided.

    For more information about these settings, see documentation for
    `sklearn.ensemble.RandomForestRegressor`.

    Args:
        scaler_type: scaler to use for training
        scaler_kwargs: keyword arguments to pass to scaler initialization
        n_jobs: number of jobs to run in parallel when training a single random forest
        random_state: random seed to use when building trees, will be
            deterministically perturbed for each training batch
        n_estimators: the number of trees in each forest
        max_depth: maximum depth of each tree, by default is unlimited
        min_samples_split: minimum number of samples required to split an internal node
        min_samples_leaf: minimum number of samples required to be at a leaf node
        max_features: number of features to consider when looking for the best split
        max_samples: if bootstrap is True, number of samples to draw
            for each base estimator
        bootstrap: whether bootstrap samples are used when building trees.
            If False, the whole dataset is used to build each tree.
    """

    scaler_type: str = "standard"
    scaler_kwargs: Optional[Mapping] = None

    # don't set default to -1 because it causes non-reproducible training
    n_jobs: int = 8
    random_state: int = 0
    n_estimators: int = 100
    max_depth: Optional[int] = None
    min_samples_split: Union[int, float] = 2
    min_samples_leaf: Union[int, float] = 1
    max_features: Union[Literal["auto", "sqrt", "log2"], int, float] = "auto"
    max_samples: Optional[Union[int, float]] = None
    bootstrap: bool = True


@dataclasses.dataclass
class _ModelTrainingConfig:
    """Convenience wrapper for model training parameters and file info

    Attrs:
        model_type: sklearn model type or keras model class to initialize
        hyperparameters: arguments to pass to model class at initialization
            time
        input_variables: variables used as features
        output_variables: variables to predict
        batch_function: name of function from `fv3fit.batches` to use for
            loading batched data
        batch_kwargs: keyword arguments to pass to batch function
        data_path: location of training data to be loaded by batch function
        scaler_type: scaler to use for training
        scaler_kwargs: keyword arguments to pass to scaler initialization
        additional_variables: list of needed variables which are not inputs
            or outputs (e.g. pressure thickness if needed for scaling)
        random_seed: value to use to initialize randomness
        validation_timesteps: timestamps to use as validation samples
        save_model_checkpoints: whether to save a copy of the model at
            each epoch
        model_path: output location for final model
        timesteps_source: one of "timesteps_file",
            "sampled_outside_input_config", "input_config", "all_mapper_times"
    """

    model_type: str
    hyperparameters: dict
    input_variables: List[str]
    output_variables: List[str]
    batch_function: str
    batch_kwargs: dict
    data_path: Optional[str] = None
    scaler_type: str = "standard"
    scaler_kwargs: dict = dataclasses.field(default_factory=dict)
    additional_variables: List[str] = dataclasses.field(default_factory=list)
    random_seed: Union[float, int] = 0
    validation_timesteps: Sequence[str] = dataclasses.field(default_factory=list)
    save_model_checkpoints: bool = False
    model_path: str = ""
    timesteps_source: str = "timesteps_file"

    def __post_init__(self):
        if self.scaler_type == "mass":
            if DELP not in self.additional_variables:
                self.additional_variables.append(DELP)

    def asdict(self):
        return dataclasses.asdict(self)

    def dump(self, path: str, filename: str = None) -> None:
        dict_ = self.asdict()
        if filename is None:
            filename = MODEL_CONFIG_FILENAME
        with fsspec.open(os.path.join(path, filename), "w") as f:
            yaml.safe_dump(dict_, f)

    @classmethod
    def load(cls, path: str) -> "_ModelTrainingConfig":
        with fsspec.open(path, "r") as f:
            config_dict = yaml.safe_load(f)
        return _ModelTrainingConfig(**config_dict)


def legacy_config_to_new_config(legacy_config: _ModelTrainingConfig) -> TrainingConfig:
    config_class = TRAINING_FUNCTIONS[legacy_config.model_type][1]
    keys = [
        "model_type",
        "hyperparameters",
        "input_variables",
        "output_variables",
        "additional_variables",
        "random_seed",
        "sample_dim_name",
    ]
    if config_class is RandomForestHyperparameters:
        for key in ("scaler_type", "scaler_kwargs"):
            legacy_config.hyperparameters[key] = getattr(legacy_config, key)
    elif config_class is DenseHyperparameters:
        legacy_config.hyperparameters[
            "save_model_checkpoints"
        ] = legacy_config.save_model_checkpoints
        fit_kwargs = legacy_config.hyperparameters.pop("fit_kwargs", {})
        if legacy_config.validation_timesteps is not None:
            fit_kwargs["validation_dataset"] = validation_dataset(legacy_config)
        legacy_config.hyperparameters["fit_kwargs"] = fit_kwargs
    else:
        raise NotImplementedError(f"unknown model type {legacy_config.model_type}")
    config_dict = dataclasses.asdict(legacy_config)
    config_dict["sample_dim_name"] = "sample"
    training_config = TrainingConfig.from_dict(
        {key: config_dict[key] for key in keys if key in config_dict}
    )
    return training_config


def load_configs(
    config_path: str,
    data_path: str,
    output_data_path: str,
    timesteps_file=None,
    validation_timesteps_file=None,
) -> Tuple[TrainingConfig, loaders.BatchesConfig, Optional[loaders.BatchesConfig]]:
    """Load training configuration information from a legacy yaml config path.

    Dumps the legacy configuration class to the output_data_path.
    """
    # TODO: remove output_data_path argument, we need it here at the moment
    # to dump legacy_config before it gets a Dataset attached to it,
    # for backwards compatibility
    # we shouldn't need this when validation_dataset is in its own data config
    # and not attached to fit_kwargs
    legacy_config = _ModelTrainingConfig.load(config_path)
    legacy_config.data_path = data_path
    legacy_config.dump(output_data_path)
    config_dict = dataclasses.asdict(legacy_config)
    training_config = legacy_config_to_new_config(legacy_config)

    data_path = config_dict["data_path"]
    batches_function = config_dict["batch_function"]
    batches_kwargs = config_dict["batch_kwargs"]

    train_batches_kwargs = {**batches_kwargs}
    if timesteps_file is not None:
        with open(timesteps_file, "r") as f:
            timesteps = yaml.safe_load(f)
        train_batches_kwargs["timesteps"] = timesteps
    train_data_config = loaders.BatchesConfig(
        data_path=data_path,
        batches_function=batches_function,
        batches_kwargs=train_batches_kwargs,
    )

    if validation_timesteps_file is not None:
        validation_batches_kwargs = {**batches_kwargs}
        with open(validation_timesteps_file, "r") as f:
            timesteps = yaml.safe_load(f)
        validation_batches_kwargs["timesteps"] = timesteps
        validation_data_config: Optional[loaders.BatchesConfig] = loaders.BatchesConfig(
            data_path=data_path,
            batches_function=batches_function,
            batches_kwargs=validation_batches_kwargs,
        )
    else:
        validation_data_config = None

    return training_config, train_data_config, validation_data_config


# TODO: this should be made to work regardless of whether we're using
# keras or sklearn models, find a way to delete this code entirely and
# use the validation BatchesConfig instead.


def check_validation_train_overlap(
    train: Sequence[str], validate: Sequence[str]
) -> None:
    overlap = set(train) & set(validate)
    if overlap:
        raise ValueError(
            f"Timestep(s) {overlap} are in both train and validation sets."
        )


def validation_timesteps_config(train_config):
    val_config = legacy_config_to_batches_config(train_config)
    assert not isinstance(val_config.data_path, list)
    val_config.batches_kwargs["timesteps"] = train_config.validation_timesteps
    val_config.batches_kwargs["timesteps_per_batch"] = len(
        train_config.validation_timesteps  # type: ignore
    )
    return val_config


# TODO: refactor all tests and code using this to create BatchesConfig
# from the beginning and delete this helper routine
def legacy_config_to_batches_config(
    legacy_config: _ModelTrainingConfig,
) -> loaders.BatchesConfig:
    return loaders.BatchesConfig(
        data_path=str(legacy_config.data_path),
        batches_function=legacy_config.batch_function,
        batches_kwargs=legacy_config.batch_kwargs,
    )


def validation_dataset(train_config: _ModelTrainingConfig,) -> Optional[xr.Dataset]:
    if len(train_config.validation_timesteps) > 0:
        check_validation_train_overlap(
            train_config.batch_kwargs["timesteps"], train_config.validation_timesteps,
        )
        validation_config = validation_timesteps_config(train_config)
        # validation config puts all data in one batch
        validation_dataset_sequence = validation_config.load_batches(
            variables=train_config.input_variables
            + train_config.output_variables
            + train_config.additional_variables
        )
        if len(validation_dataset_sequence) > 1:
            raise ValueError(
                "Something went wrong! "
                "All validation data should be concatenated into a single batch. "
                f"There are {len(validation_dataset_sequence)} "
                "batches in the sequence."
            )
        return validation_dataset_sequence[0]
    else:
        validation_dataset = None
    return validation_dataset


def load_training_config(model_path: str) -> _ModelTrainingConfig:
    """Load training configuration information from a model directory URL.
    Note:
        This loads a file that you would get from using ModelTrainingConfig.dump
        with no filename argument, as is done by fv3fit.train. To ensure
        backwards compatibility, you should use this routine to load such
        a file instead of manually specifying the filename.
        The default filename may change in the future.
    Args:
        model_path: model dir dumped by fv3fit.dump
    Returns:
        dict: training config dict
    """
    config_path = os.path.join(model_path, MODEL_CONFIG_FILENAME)
    return _ModelTrainingConfig.load(config_path)
