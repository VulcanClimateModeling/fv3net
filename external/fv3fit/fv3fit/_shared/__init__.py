from .config import (
    TrainingConfig,
    register_training_function,
)
from .packer import (
    pack,
    unpack,
    count_features,
    ArrayPacker,
    unpack_matrix,
    multiindex_to_tuple,
    tuple_to_multiindex,
)
from .scaler import (
    StandardScaler,
    ManualScaler,
    get_mass_scaler,
    get_scaler,
    NormalizeTransform,
)
from .predictor import Predictor
from .stacking import (
    StackedBatches,
    stack_non_vertical,
    stack,
    match_prediction_to_input_coords,
    SAMPLE_DIM_NAME,
)
from .models import EnsembleModel, DerivedModel
from .filesystem import get_dir, put_dir
