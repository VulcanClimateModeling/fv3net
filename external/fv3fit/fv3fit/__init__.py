from ._shared.predictor import Predictor, Estimator
from ._shared.io import dump, load
from . import keras
from . import sklearn