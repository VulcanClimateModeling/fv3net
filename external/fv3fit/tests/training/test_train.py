import dataclasses
from typing import Any, Callable, Optional, Sequence, TextIO, Tuple

from fv3fit._shared.config import SliceConfig
from fv3fit.keras._models.shared.clip import ClipConfig
from fv3fit.keras._models.convolutional import ConvolutionalHyperparameters
from fv3fit.keras._models.shared.convolutional_network import ConvolutionalNetworkConfig
from fv3fit.keras._models.shared.pure_keras import PureKerasModel
import pytest
import xarray as xr
import numpy as np
import fv3fit
from fv3fit._shared.config import TRAINING_FUNCTIONS, get_hyperparameter_class
import vcm.testing
import tempfile
from fv3fit.keras._models.precipitative import LV, CPD, GRAVITY
from fv3fit._shared.stacking import stack, stack_non_vertical, SAMPLE_DIM_NAME
from fv3fit.train import tfdataset_from_batches
import tensorflow as tf


# training functions that work on arbitrary datasets, can be used in generic tests below
GENERAL_TRAINING_TYPES = [
    "convolutional",
    "sklearn_random_forest",
    "precipitative",
    "dense",
    "transformed",
]
# training functions that have restrictions on the datasets they support,
# cannot be used in generic tests below
# you must write a separate file that specializes each of the tests
# for models in this list
SPECIAL_TRAINING_TYPES = []


# automatically test on every registered training class
@pytest.fixture(params=GENERAL_TRAINING_TYPES)
def model_type(request):
    return request.param


def test_all_training_functions_are_tested_or_exempted():
    missing_types = set(TRAINING_FUNCTIONS.keys()).difference(
        GENERAL_TRAINING_TYPES + SPECIAL_TRAINING_TYPES
    )
    assert len(missing_types) == 0, (
        "training type must be added to GENERAL_TRAINING_TYPES or "
        "SPECIAL_TRAINING_TYPES in test script"
    )


SYSTEM_DEPENDENT_TYPES = [
    "convolutional",
    "sklearn_random_forest",
    "precipitative",
    "dense",
]
"""model types which produce different results on different systems"""


def test_training_functions_exist():
    assert len(TRAINING_FUNCTIONS.keys()) > 0


@dataclasses.dataclass
class TrainingResult:
    model: fv3fit.Predictor
    output_variables: Sequence[str]
    test_dataset: xr.Dataset
    hyperparameters: Any


def get_default_hyperparameters(model_type, input_variables, output_variables):
    """
    Returns a hyperparameter configuration class for the model type with default
    values.
    """
    cls = get_hyperparameter_class(model_type)

    try:
        return cls.get_default_model(input_variables, output_variables)
    except AttributeError:
        pass

    try:
        hyperparameters = cls()
    except TypeError:
        hyperparameters = cls(
            input_variables=input_variables, output_variables=output_variables
        )
    return hyperparameters


def train_identity_model(model_type, sample_func, hyperparameters=None):
    input_variables, output_variables, train_dataset = get_dataset(
        model_type, sample_func
    )
    if hyperparameters is None:
        hyperparameters = get_default_hyperparameters(
            model_type, input_variables, output_variables
        )
    input_variables, output_variables, test_dataset = get_dataset(
        model_type, sample_func
    )
    train_tfdataset = tfdataset_from_batches([train_dataset for _ in range(10)])
    val_tfdataset = tfdataset_from_batches([test_dataset])
    train = fv3fit.get_training_function(model_type)
    model = train(hyperparameters, train_tfdataset, val_tfdataset)
    test_dataset = unstack_test_dataset(test_dataset)
    return TrainingResult(model, output_variables, test_dataset, hyperparameters)


def unstack_test_dataset(test_dataset: xr.Dataset):
    """
    Undo the stacking used to create stacked tf.data.Datasets,
    so we can call .predict on unstacked data.
    """
    return test_dataset.unstack(SAMPLE_DIM_NAME).transpose(
        "sample", "tile", "x", "y", "z"
    )


def _process(input_variables, input_values, output_variables, output_values):
    data_vars = {name: value for name, value in zip(input_variables, input_values)}
    data_vars[input_variables[0]] = data_vars[input_variables[0]].astype(np.float32)
    data_vars.update(
        {name: value for name, value in zip(output_variables, output_values)}
    )
    return xr.Dataset(data_vars=data_vars)


def _get_dataset_precipitative(sample_func):
    input_variables = [
        "air_temperature",
        "specific_humidity",
        "pressure_thickness_of_atmospheric_layer",
        "physics_precip",
    ]
    output_variables = ["dQ1", "dQ2", "total_precipitation_rate"]
    input_values = list(sample_func() for _ in input_variables)
    i_phys_prec = input_variables.index("physics_precip")
    input_values[i_phys_prec] = input_values[i_phys_prec].isel(z=0) * 0.0
    output_values = (
        input_values[0] - LV / CPD * input_values[1],  # latent heat of condensation
        input_values[1],
        1.0
        / GRAVITY
        * np.sum(
            input_values[2] * input_values[1], axis=-1
        ),  # total_precipitation_rate is integration of dQ2
    )
    return (
        input_variables,
        output_variables,
        _process(input_variables, input_values, output_variables, output_values),
    )


def _get_dataset_default(sample_func, data_2d_ceof=1):
    input_variables = ["var_in_2d", "var_in_3d"]  # 2d var will be clipped below
    output_variables = ["var_out"]
    input_values = list(sample_func() for _ in input_variables)
    i_2d_input = input_variables.index("var_in_2d")
    input_values[i_2d_input] = input_values[i_2d_input].isel(z=0) * data_2d_ceof
    i_3d_input = input_variables.index("var_in_3d")
    output_values = [input_values[i_3d_input]]
    return (
        input_variables,
        output_variables,
        _process(input_variables, input_values, output_variables, output_values),
    )


def get_dataset_default(sample_func):
    input_variables, output_variables, train_dataset = _get_dataset_default(sample_func)
    train_dataset = stack_non_vertical(train_dataset)
    return input_variables, output_variables, train_dataset


def get_dataset_precipitative(sample_func):
    input_variables, output_variables, train_dataset = _get_dataset_precipitative(
        sample_func
    )
    train_dataset = stack_non_vertical(train_dataset)
    return input_variables, output_variables, train_dataset


def get_dataset_convolutional(sample_func):
    input_variables, output_variables, train_dataset = _get_dataset_default(
        sample_func,
        # transpose tests fail if the 2d data varies
        data_2d_ceof=0.0,
    )
    train_dataset = stack(train_dataset, unstacked_dims=["tile", "x", "y", "z"])
    return input_variables, output_variables, train_dataset


def get_dataset(
    model_type, sample_func
) -> Tuple[Sequence[str], Sequence[str], tf.data.Dataset]:

    build_test_case = globals().get(f"get_dataset_{model_type}", get_dataset_default)
    return build_test_case(sample_func)


def assert_can_learn_identity(
    model_type,
    sample_func: Callable[[], xr.DataArray],
    max_rmse: float,
    regtest: Optional[TextIO] = None,
):
    """
    Args:
        model_type: type of model to train
        hyperparameters: model configuration
        sample_func: function that returns example DataArrays for training and
            validation, should return different data on subsequent calls
        max_rmse: maximum permissible root mean squared error
        regtest: if given, write hash of output dataset to this file object
    """
    result = train_identity_model(model_type, sample_func=sample_func)
    out_dataset = result.model.predict(result.test_dataset)
    for name in result.output_variables:
        assert out_dataset[name].dims == result.test_dataset[name].dims
    rmse = (
        np.mean(
            [
                np.mean((out_dataset[name] - result.test_dataset[name]) ** 2)
                / np.std(result.test_dataset[name]) ** 2
                for name in result.output_variables
            ]
        )
        ** 0.5
    )
    assert rmse < max_rmse
    if model_type in SYSTEM_DEPENDENT_TYPES:
        print(f"{model_type} is system dependent, not checking against regtest output")
    else:
        for result in vcm.testing.checksum_dataarray_mapping(result.test_dataset):
            print(result, file=regtest)
        for result in vcm.testing.checksum_dataarray_mapping(out_dataset):
            print(result, file=regtest)


@pytest.mark.slow
def test_train_default_model_on_identity(model_type, regtest):
    """
    The model with default configuration options can learn the identity function,
    using gaussian-sampled data around 0 with unit variance.
    """
    fv3fit.set_random_seed(1)
    # don't set n_feature too high for this, because of curse of dimensionality
    n_sample, n_tile, nx, ny, n_feature = 5, 6, 12, 12, 2
    sample_func = get_uniform_sample_func(size=(n_sample, n_tile, nx, ny, n_feature))

    assert_can_learn_identity(
        model_type, sample_func=sample_func, max_rmse=0.2, regtest=regtest,
    )


@pytest.mark.slow
def test_default_convolutional_model_is_transpose_invariant():
    """
    The model with default configuration options can learn the identity function,
    using gaussian-sampled data around 0 with unit variance.
    """
    fv3fit.set_random_seed(1)
    # don't set n_feature too high for this, because of curse of dimensionality
    n_sample, n_tile, nx, ny, n_feature = 10, 6, 12, 12, 2
    sample_func = get_uniform_sample_func(size=(n_sample, n_tile, nx, ny, n_feature))
    result = train_identity_model("convolutional", sample_func=sample_func)
    transpose_input = result.test_dataset.copy(deep=True)
    transpose_input["var_in_3d"].values[:] = np.transpose(
        transpose_input["var_in_3d"].values, axes=(0, 1, 3, 2, 4)
    )
    transpose_output = result.model.predict(result.test_dataset)
    transpose_output["var_out"].values[:] = np.transpose(
        transpose_output["var_out"].values, axes=(0, 1, 3, 2, 4)
    )
    output_from_transpose = result.model.predict(transpose_input)
    n_halo = result.hyperparameters.convolutional_network.halos_required
    # transposing tile data messes up neighbors, so we have to assess only on
    # data that has no halo dependence
    assert n_halo * 2 < nx
    assert n_halo * 2 < ny
    xr.testing.assert_allclose(
        output_from_transpose.isel(x=slice(n_halo, -n_halo), y=slice(n_halo, -n_halo)),
        transpose_output.isel(x=slice(n_halo, -n_halo), y=slice(n_halo, -n_halo)),
        atol=1e-5,
    )


@pytest.mark.slow
def test_diffusive_convolutional_model_gives_bounded_output():
    """
    The model with diffusive enabled should give outputs in the range of the inputs.
    """
    fv3fit.set_random_seed(1)
    # don't set n_feature too high for this, because of curse of dimensionality
    n_sample, n_tile, nx, ny, n_feature = 10, 6, 12, 12, 2
    low, high = 0.0, 1.0
    train_sample_func = get_uniform_sample_func(
        size=(n_sample, n_tile, nx, ny, n_feature), low=low - 1.0, high=high + 1.0
    )
    test_sample_func = get_uniform_sample_func(
        size=(n_sample, n_tile, nx, ny, n_feature), low=low, high=high
    )
    input_variables, output_variables, test_dataset = get_dataset(
        model_type="convolutional", sample_func=test_sample_func
    )
    hyperparameters = ConvolutionalHyperparameters(
        input_variables=input_variables,
        output_variables=output_variables,
        convolutional_network=ConvolutionalNetworkConfig(diffusive=True),
    )
    result = train_identity_model(
        "convolutional", sample_func=train_sample_func, hyperparameters=hyperparameters
    )
    output = result.model.predict(unstack_test_dataset(test_dataset))
    for name, data_array in output.data_vars.items():
        assert np.all(data_array.values >= low), name
        assert np.all(data_array.values <= high), name


@pytest.mark.slow
def test_train_with_same_seed_gives_same_result(model_type):
    n_sample, n_tile, nx, ny, n_feature = 1, 6, 12, 12, 2
    fv3fit.set_random_seed(0)

    sample_func = get_uniform_sample_func(size=(n_sample, n_tile, nx, ny, n_feature))
    first_result = train_identity_model(model_type, sample_func)
    fv3fit.set_random_seed(0)
    sample_func = get_uniform_sample_func(size=(n_sample, n_tile, nx, ny, n_feature))
    second_result = train_identity_model(model_type, sample_func)
    xr.testing.assert_equal(first_result.test_dataset, second_result.test_dataset)
    first_output = first_result.model.predict(first_result.test_dataset)
    second_output = second_result.model.predict(first_result.test_dataset)
    xr.testing.assert_equal(first_output, second_output)


@pytest.mark.slow
def test_predict_does_not_mutate_input(model_type):
    n_sample, n_tile, nx, ny, n_feature = 1, 6, 12, 12, 2
    sample_func = get_uniform_sample_func(size=(n_sample, n_tile, nx, ny, n_feature))
    result = train_identity_model(model_type, sample_func=sample_func)
    hash_before_predict = vcm.testing.checksum_dataarray_mapping(result.test_dataset)
    _ = result.model.predict(result.test_dataset)
    assert (
        vcm.testing.checksum_dataarray_mapping(result.test_dataset)
        == hash_before_predict
    ), "predict should not mutate its input"


def get_uniform_sample_func(size, low=0, high=1, seed=0):
    random = np.random.RandomState(seed=seed)

    def sample_func():
        return xr.DataArray(
            random.uniform(low=low, high=high, size=size),
            dims=["sample", "tile", "x", "y", "z"],
            coords=[range(size[i]) for i in range(len(size))],
        )

    return sample_func


@pytest.mark.slow
def test_train_default_model_on_nonstandard_identity(model_type):
    """
    The model with default configuration options can learn the identity function,
    using gaussian-sampled data around a non-zero value with non-unit variance.
    """
    low, high = 100, 200
    # don't set n_feature too high for this, because of curse of dimensionality
    n_sample, n_tile, nx, ny, n_feature = 5, 6, 12, 12, 2
    sample_func = get_uniform_sample_func(
        low=low, high=high, size=(n_sample, n_tile, nx, ny, n_feature)
    )

    assert_can_learn_identity(
        model_type, sample_func=sample_func, max_rmse=0.2 * (high - low),
    )


def remove_key(d: dict, key: Any):
    """
    Removes the given key from dict and all subdicts, if present.
    """
    return_dict = {**d}
    return_dict.pop(key, None)
    for name, value in return_dict.items():
        if isinstance(value, dict):
            return_dict[name] = remove_key(value, key)
    return return_dict


@pytest.mark.slow
def test_dump_and_load_default_maintains_prediction(model_type):
    n_sample, n_tile, nx, ny, n_feature = 1, 6, 12, 12, 2
    sample_func = get_uniform_sample_func(size=(n_sample, n_tile, nx, ny, n_feature))
    result = train_identity_model(model_type, sample_func=sample_func)

    original_result = result.model.predict(result.test_dataset)
    with tempfile.TemporaryDirectory() as tmpdir:
        fv3fit.dump(result.model, tmpdir)
        loaded_model = fv3fit.load(tmpdir)
    if isinstance(result.model, PureKerasModel):
        # we used to test for bit-identical results, but the convolutional model is
        # somehow not bit-identical when reloaded, so we've added these checks that at
        # least the model weights, config, and summary are bit-identical.
        assert result.model.model.summary() == loaded_model.model.summary()
        for l1, l2 in zip(result.model.model.layers, loaded_model.model.layers):
            loaded_config = remove_key(l2.get_config(), "shared_object_id")
            assert l1.get_config() == loaded_config
        assert all(
            [
                np.all(w1 == w2)
                for w1, w2 in zip(
                    result.model.model.weights, loaded_model.model.weights
                )
            ]
        )
    loaded_result = loaded_model.predict(result.test_dataset)
    xr.testing.assert_allclose(loaded_result, original_result, rtol=1e-4)


@pytest.mark.slow
def test_train_dense_model_clipped_inputs_outputs():
    da = xr.DataArray(np.arange(1500).reshape(150, 10) * 1.0, dims=["sample", "z"],)
    train_dataset = xr.Dataset(
        data_vars={"var_in_0": da, "var_in_1": da, "var_out_0": da, "var_out_1": da}
    )
    train_batches = [train_dataset for _ in range(2)]
    val_batches = train_batches
    train = fv3fit.get_training_function("dense")

    input_variables = ["var_in_0", "var_in_1"]
    output_variables = ["var_out_0", "var_out_1"]

    hyperparameters = get_default_hyperparameters(
        "dense", input_variables, output_variables
    )
    hyperparameters.clip_config = ClipConfig(
        {"var_in_0": SliceConfig(2, 5), "var_out_0": SliceConfig(4, 8)}
    )
    model = train(
        hyperparameters,
        tfdataset_from_batches(train_batches),
        tfdataset_from_batches(val_batches),
    )
    prediction = model.predict(train_dataset)
    assert np.unique(prediction["var_out_0"].isel(z=slice(None, 4)).values) == 0.0
    assert np.unique(prediction["var_out_0"].isel(z=slice(8, None)).values) == 0.0
