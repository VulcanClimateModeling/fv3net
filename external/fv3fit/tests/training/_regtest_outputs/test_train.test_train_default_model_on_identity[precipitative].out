('air_temperature', '7b32e49167f2a99def51bb0d7733e67b')
('dQ1', '20ffc97e55f5596a8b1660617fdf2556')
('dQ2', 'a4a21f03f8e01053cfd28025fefe1cde')
('pressure_thickness_of_atmospheric_layer', '2641737e6bd80967917eafa3340a7232')
('specific_humidity', 'a4a21f03f8e01053cfd28025fefe1cde')
('total_precipitation_rate', 'd17554236642810ffa9f1215dec61b1f')
('dQ1', '7bd29b14ef1ecfc56ad4e8a50cc017d2')
('dQ2', '9e8aa785fafce6e8b69319795428507f')
('total_precipitation_rate', '182b0cedb4238d6436049a687b1c40ef')
